src/render.c

```diff
@@ -229,21 +229,106 @@ uint32_t make_rectangle(int x, int y, int wid, int hei, xcb_render_trapezoid_t t
 	return 1;
 }
 
-void render(session_t *ps, int x, int y, int dx, int dy, int wid, int hei, double opacity,
-            bool argb, bool neg, xcb_render_picture_t pict, glx_texture_t *ptex,
-            const region_t *reg_paint, const glx_prog_main_t *pprogram) {
+uint32_t make_rounded_window_shape(xcb_render_trapezoid_t traps[], uint32_t max_ntraps,
+                                   int cr, int wid, int hei) {
+	uint32_t n = make_circle(cr, cr, cr, max_ntraps, traps);
+	n += make_circle(wid - cr, cr, cr, max_ntraps, traps + n);
+	n += make_circle(wid - cr, hei - cr, cr, max_ntraps, traps + n);
+	n += make_circle(cr, hei - cr, cr, max_ntraps, traps + n);
+	n += make_rectangle(0, cr, cr, hei - 2 * cr, traps + n);
+	n += make_rectangle(cr, 0, wid - 2 * cr, cr, traps + n);
+	n += make_rectangle(wid - cr, cr, cr, hei - 2 * cr, traps + n);
+	n += make_rectangle(cr, hei - cr, wid - 2 * cr, cr, traps + n);
+	n += make_rectangle(cr, cr, wid - 2 * cr, hei - 2 * cr, traps + n);
```

Comment:
Can't we reduce this to 4 circles and 3 rectangles?
```diff
diff --git a/src/render.c b/src/render.c
index a328aff..a4e536f 100644
--- a/src/render.c
+++ b/src/render.c
@@ -235,11 +235,9 @@ uint32_t make_rounded_window_shape(xcb_render_trapezoid_t traps[], uint32_t max_
        n += make_circle(wid - cr, cr, cr, max_ntraps, traps + n);
        n += make_circle(wid - cr, hei - cr, cr, max_ntraps, traps + n);
        n += make_circle(cr, hei - cr, cr, max_ntraps, traps + n);
-       n += make_rectangle(0, cr, cr, hei - 2 * cr, traps + n);
+       n += make_rectangle(0, cr, wid, hei - 2 * cr, traps + n);
        n += make_rectangle(cr, 0, wid - 2 * cr, cr, traps + n);
-       n += make_rectangle(wid - cr, cr, cr, hei - 2 * cr, traps + n);
        n += make_rectangle(cr, hei - cr, wid - 2 * cr, cr, traps + n);
-       n += make_rectangle(cr, cr, wid - 2 * cr, hei - 2 * cr, traps + n);
        return n;
 }

@@ -268,7 +266,7 @@ void render(session_t *ps, int x, int y, int dx, int dy, int wid, int hei, int f
                                                           p_tmp, trans, 1, &rect);

                                uint32_t max_ntraps = to_u32_checked(cr);
-                               xcb_render_trapezoid_t traps[4 * max_ntraps + 5];
+                               xcb_render_trapezoid_t traps[4 * max_ntraps + 3];

                                uint32_t n = make_rounded_window_shape(
                                    traps, max_ntraps, cr, fullwid, fullhei);
@@ -747,7 +745,7 @@ win_paint_shadow(session_t *ps, struct managed_win *w, region_t *reg_paint) {
        xcb_render_picture_t td = XCB_NONE;
        if (w->corner_radius) {
                uint32_t max_ntraps = to_u32_checked(w->corner_radius);
-               xcb_render_trapezoid_t traps[4 * max_ntraps + 5];
+               xcb_render_trapezoid_t traps[4 * max_ntraps + 3];
                uint32_t n = make_rounded_window_shape(
                    traps, max_ntraps, w->corner_radius, w->widthb, w->heightb);

@@ -897,7 +895,7 @@ win_blur_background(session_t *ps, struct managed_win *w, xcb_render_picture_t t
                xcb_render_picture_t td = XCB_NONE;
                if (cr) {
                        uint32_t max_ntraps = to_u32_checked(cr);
-                       xcb_render_trapezoid_t traps[4 * max_ntraps + 5];
+                       xcb_render_trapezoid_t traps[4 * max_ntraps + 3];
                        uint32_t n =
                            make_rounded_window_shape(traps, max_ntraps, cr, wid, hei);

```

Reply:
Yep